File: frontend/src/constants/taxa.ts
import type { ActiveTaxon } from '../contexts/activeTaxaContext';
import type { TaxonomicRank } from '../types/taxa';

export type ItalicizedRank = Extract<
    TaxonomicRank,
    'genus' | 'species' | 'subspecies' | 'subgenus'
>;

export const italicizedRanks: ItalicizedRank[] = [
    'genus',
    'species',
    'subspecies',
    'subgenus',
];

export const TAXON_COLORS = [
    '#e58606',
    '#5d69b1',
    '#52bca3',
    '#99c945',
    '#cc61b0',
    '#24796c',
    '#daa51b',
    '#2f8ac4',
    '#764e9f',
    '#ed645a',
    '#cc3a8e',
    '#a5aa99',
];

export const DUMMY_TAXON: ActiveTaxon = {
    color: TAXON_COLORS[0],
    taxonLoading: false,
    taxonError: false,
    dateCounts: null,
    dateCountsLoading: false,
    dateRangeLoading: false,
    datasetCountsLoading: false,
    nSValuesLoading: false,
    lastLoadedID: 'G9Q6',
    taxonID: 'G9Q6',
    info: {
        taxonID: 'G9Q6',
        acceptedNameUsageID: 'G9Q6',
        parentNameUsageID: '7NQXN',
        scientificName: 'Archispirostreptus gigas',
        canonicalName: 'Archispirostreptus gigas',
        taxonRank: 'species',
        scientificNameAuthorship: '(Peters, 1855)',
        kingdom: 'Animalia',
        phylum: 'Arthropoda',
        class: 'Diplopoda',
        order: 'Spirostreptida',
        family: 'Spirostreptidae',
        genericName: 'Archispirostreptus',
        specificEpithet: 'gigas',
        infragenericEpithet: null,
        infraspecificEpithet: null,
        uSInvasive: false,
        taxonomicStatus: 'accepted',
        vernacularNames: ['Giant African Millipede', 'Shongololo'],
        nSRankState: null,
        nSRankStateNoINat: null,
    },
    nSValues: {
        numberOfOccurrences: 40,
        rangeExtentKm2: 8093446,
        areaOfOccupancy1Km2Bins: 40,
        areaOfOccupancy4Km2Bins: 39,
        observationCount: 41,
    },
    datasetCounts: { 'iNaturalist Research-grade Observations': 15 },
    dateMin: null,
    dateMax: null,
};

File: backend/models/taxa.py
# Taxon related models
from typing import List

from pydantic import BaseModel, ConfigDict, Field, alias_generators
from backend.constants.taxa import TaxonomicRank


class TaxonSuggestion(BaseModel):
    model_config = ConfigDict(
        alias_generator=alias_generators.to_camel,
        populate_by_name=True,
    )
    canonical_name: str | None
    scientific_name_authorship: str | None
    taxon_id: str = Field(alias='taxonID')
    taxon_rank: str | None
    us_invasive: bool | None = Field(alias='uSInvasive')
    taxonomic_status: str | None


class TaxonInfo(BaseModel):
    model_config = ConfigDict(
        alias_generator=alias_generators.to_camel,
        populate_by_name=True,
    )
    # Fit convention of acronym caps
    taxon_id: str = Field(alias='taxonID')
    scientific_name_authorship: str | None
    vernacular_names: List[str] | None
    # Fit convention of acronym caps
    accepted_name_usage_id: str | None = Field(
        None, alias='acceptedNameUsageID')
    # Fit convention of acronym caps
    parent_name_usage_id: str | None = Field(None, alias='parentNameUsageID')
    canonical_name: str | None
    scientific_name: str | None
    taxon_rank: TaxonomicRank | None
    us_invasive: bool | None = Field(None, alias='uSInvasive')
    taxonomic_status: str | None
    ns_rank_state: str | None = Field(None, alias='nSRankState')
    ns_rank_state_no_inat: str | None = Field(None, alias='nSRankStateNoINat')
    kingdom: str | None
    phylum: str | None
    # Aliased for reserved word 'class'
    taxon_class: str | None = Field(None, alias='class')
    order: str | None
    family: str | None
    generic_name: str | None
    infrageneric_epithet: str | None
    specific_epithet: str | None
    infraspecific_epithet: str | None
